### PR TITLE
Integrate Java Lab with Code Studio

### PR DESCRIPTION
**Pull request by molly-moen**
Plumbing for integrating Java Lab with code studio. Heavily based on similar work for [ai lab](https://github.com/code-dot-org/code-dot-org/pull/37495). 

This work enables levelbuilders to access the Java Lab by going to `studio.code.org/projects/javalab/new`. This will spin up a new project that can be renamed. Any code put in the editor is not saved yet. It removes the previous url of `studio.code.org/javalab`. The project will be labeled as "Java Lab" on the my projects page.

The goals of this PR are:
- enable projects url for Java Lab
- keep Java Lab hidden for all but levelbuilders
- get the lab ready for further integration with projects and levelbuilder

I recommend viewing this PR with whitespace changes hidden.

This is what spinning up a new JavaLab project looks like:
Screen Shot 2021-02-17 at 12 44 40 PM

## Links
- [jira ticket](https://codedotorg.atlassian.net/browse/CSA-119)

## Testing story
Tested locally with both levelbuilder and non-levelbuilder permissions (to ensure only levelbuilder permissions allow viewing the lab).

## Follow-up work
- Save code written in editor (includes setting up file naming)
- Integrate with levelbuilder, including:
   - Integrate instructions on editor with level definition
   - Make bubble progression work (not tested since we just have a project currently)
- Make the UI more accessible 
   

## PR Checklist:

- [ ] Tests provide adequate coverage
- [ ] Privacy and Security impacts have been assessed
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

File: apps/src/javalab/Javalab.js [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+import {Provider} from 'react-redux';
+import {getStore, registerReducers} from '../redux';
+import JavalabView from './JavalabView';
+import javalab from './javalabRedux';
+import {TestResults} from '@cdo/apps/constants';
+
+/**
+ * On small mobile devices, when in portrait orientation, we show an overlay
+ * image telling the user to rotate their device to landscape mode.  Because
+ * the ailab app is able to render at a minimum width of 480px, we set this
```
Comment (jmkulwik) [inline apps/src/javalab/Javalab.js]: NIT: Reference to AI Lab

File: apps/src/javalab/Javalab.js [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+  this.level = null;
+
+  /** @type {StudioApp} */
+  this.studioApp_ = null;
+};
+
+/**
+ * Inject the studioApp singleton.
+ */
+Javalab.prototype.injectStudioApp = function(studioApp) {
```
Comment (jmkulwik) [inline apps/src/javalab/Javalab.js:30]: We haven't been consistent on this everywhere in other apps, but this should probably be JavaLab and JavaLab.js (similar to GameLab and SpriteLab). Reasoning: On the product side, we'll always refer to it as two separate words "Java Lab" so camelcasing the two words would make sense.

Comment (jmkulwik) [reply apps/src/javalab/Javalab.js:30]: Huh now I'm noticing we're _very_ inconsistent about this in all our apps.

File: apps/src/javalab/Javalab.js (hunk 1 above)
Comment (molly-moen) [reply apps/src/javalab/Javalab.js]: 🤦

File: apps/src/javalab/Javalab.js (hunk 2 above)
Comment (molly-moen) [reply apps/src/javalab/Javalab.js:30]: yeah I'm happy to go with either (camel casing probably makes more sense?) but it seemed like keeping lab lowercase was the norm in general

File: dashboard/app/controllers/projects_controller.rb [hunk 3]
```diff
@@ -138,6 +138,10 @@ class ProjectsController < ApplicationController
     },
     eval: {
       name: 'Eval Free Play'
+    },
+    javalab: {
+      name: 'New Java Lab Project',
+      levelbuilder_required: true
```
Comment (jmkulwik) [inline dashboard/app/controllers/projects_controller.rb:144]: TIL about `levelbuilder_required: true` 🤯

File: apps/src/javalab/Javalab.js (hunk 2 above)
Comment (jmkulwik) [reply apps/src/javalab/Javalab.js:30]: Well. haha. When I first read this, I felt strongly that camel casing was the right option. Then I saw what we do everywhere else. I'm happy with whatever you choose.

**Review by jmkulwik — APPROVED** [review]
Nice work finding all this!

File: apps/src/javalab/Javalab.js [hunk 4]
```diff
@@ -0,0 +1,125 @@
+import React from 'react';
+import ReactDOM from 'react-dom';
+import {Provider} from 'react-redux';
+import {getStore, registerReducers} from '../redux';
+import JavalabView from './JavalabView';
+import javalab from './javalabRedux';
+import {TestResults} from '@cdo/apps/constants';
+
+/**
+ * On small mobile devices, when in portrait orientation, we show an overlay
```
Comment (ajpal) [inline apps/src/javalab/Javalab.js:10]: Is Java Lab going to be supported on mobile/tablet?

**Comment by ajpal** [comment]
Is Java Lab going to be the product name for this environment as well? Mostly curious because this is the first lab that will be named for its language rather than its intended use (Web Lab not HTML Lab, App Lab not JS Lab, etc). I don't have an alternative to propose or anything, just curious if it's something you've thought about

File: apps/src/javalab/Javalab.js (hunk 4 above)
Comment (molly-moen) [reply apps/src/javalab/Javalab.js:10]: tablet yes, mobile not in pilot and I think only view-only.

**Comment by molly-moen** [comment]
> Is Java Lab going to be the product name for this environment as well? Mostly curious because this is the first lab that will be named for its language rather than its intended use (Web Lab not HTML Lab, App Lab not JS Lab, etc). I don't have an alternative to propose or anything, just curious if it's something you've thought about

@ajpal, yes we talked about various possibilities for naming, and since Java is the defining feature of this environment, we went with Java Lab (the only other defining feature is it's an IDE, but you could argue our other labs are also IDEs). I think it's our plan to use that on the product side as well but it could possibly change.

**Pull request merged**